perun/logic/stats.py

```diff
@@ -341,3 +482,184 @@ def _modify_stats_file(stats_filepath, stats_ids, stats_contents, modify_functio
+    of custom directories or files that were not created by the stats interface.
+
+    :return tuple: list of minor versions (checksum, date),
+                   list of custom directories and files
+    """
+
+    def dirs_generator(directory, custom_list, filter_func=None):
+        """ Generator of directories contained within the 'directory'. Files or objects not passing
+        the filter function are appended to the custom list.
+
+        :param str directory: path of the base directory to scan for other directories
+        :param list custom_list: list for custom objects that are not valid directories
+        :param function filter_func: optional function that can filter the traversed directories
+
+        :return generator: generator object that provides the valid directories
+        """
+        for item in os.listdir(directory):
+            item = os.path.join(directory, item)
+            # Filter out objects that are not directories or do not pass the filtering function
+            if not os.path.isdir(item) or (filter_func is not None and not filter_func(item)):
+                custom_list.append(item)
+            # The rest should be valid directories
+            else:
+                yield item
+
+    # List all the upper and lower directories
+    versions, custom = [], []
+    stats_dir, stats_idx = pcs.get_stats_directory(), pcs.get_stats_index()
+    # The upper level of directories should represent the first SHA byte
+    for upper in list(dirs_generator(stats_dir, custom, os.listdir)):
+        # The lower level represents the rest of the SHA
+        lower_list = list(dirs_generator(upper, custom))
+        # If there are no lower level directories, then the upper directory is custom
+        if not lower_list:
+            custom.append(upper)
+        # Check all the lower level objects
+        temp_versions, temp_custom = [], []
+        for lower in lower_list:
+            try:
+                # Construct the minor version from SHA and resolve it
+                temp_versions.append(_get_version_info(store.sha_path_to_sha(lower)))
+                # Check the contents of the minor version stats, directories should not be allowed
+                # Do not check the files as we do not really have a way to distinguish custom ones
+                temp_custom.extend(list(dirs_generator(lower, [])))
+            except exceptions.VersionControlSystemException:
+                temp_custom.append(lower)
+        # If all the objects in the upper directory are custom, delete the upper
+        # Otherwise delete only the lower level objects
+        if not temp_versions:
+            custom.append(upper)
+        else:
+            versions.extend(temp_versions)
+            custom.extend(temp_custom)
+
+    # Remove the .index file from the custom list if it is present
+    if stats_idx in custom:
+        custom.remove(stats_idx)
+
+    return versions, custom
+
+
+def _index_loader_wrapper():
```

Comment:
Suggestion: Maybe rename to something more function-like?

Reply:
Renamed to a '_load_stats_index'